include/boost/spirit/home/x3/directive/with.hpp

```diff
@@ -15,21 +15,40 @@ namespace boost { namespace spirit { namespace x3
     ///////////////////////////////////////////////////////////////////////////
     // with directive injects a value into the context prior to parsing.
     ///////////////////////////////////////////////////////////////////////////
+    template <typename Subject, typename Derived, typename T>
+    struct with_value_holder
+      : unary_parser<Subject, Derived>
+    {
+        typedef unary_parser<Subject, Derived> base_type;
+        mutable T val;
```

Comment:
Could you explain why is it `mutable`?

1. The tests you had added do not cover it.
2. Allowing to mutate the value creates thread-safety risks.
3. It does not play well with `constexpr`, especially with Clang 6 and below.

Reply:
The idea was rvalue injection should be mutable in const parser as well. Given parsers are immutable in the parse functions, rvalue injection would be immutable without `mutable` keyword. 

But I also agree with you on the issues it brings.  I don't have a solution at the moment.